cardano-cli/src/Cardano/CLI/Slot.hs

```diff
@@ -0,0 +1,329 @@
+  , SyncProgress(..)
+  , SyncTolerance(..)
+  , TimeInterpreter(..)
+
+  , getSyncProgress
+  , mkTimeInterpreter
+  , neverFails
+  ) where
+
+import           Prelude
+
+import           Cardano.Slotting.Slot (EpochSize (..), SlotNo)
+import           Control.DeepSeq (NFData)
+import           Control.Exception (throwIO)
+import           Control.Monad ((>=>))
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Control.Monad.Trans.Except (ExceptT (..), runExceptT)
+import           Control.Monad.Trans.Maybe (MaybeT (..), exceptToMaybeT)
+import           Data.ByteString (ByteString)
+import           Data.Either (fromRight)
+import           Data.Hashable (Hashable)
+import           Data.Kind (Type)
+import           Data.Maybe (fromMaybe)
+import           Data.Ratio ((%))
+import           Data.Time.Clock (NominalDiffTime, UTCTime, getCurrentTime)
+import           Data.Word (Word32)
+import           GHC.Generics (Generic)
+import           GHC.Stack (HasCallStack)
+import           GHC.TypeLits (Symbol)
+import           Ouroboros.Consensus.BlockchainTime.WallClock.Types (RelativeTime (..),
+                   SystemStart (..), diffRelTime)
+import           Ouroboros.Consensus.HardFork.History.Qry (Interpreter, PastHorizonException (..),
+                   mkInterpreter, slotToWallclock)
+import           Ouroboros.Consensus.HardFork.History.Summary (neverForksSummary)
+import           Quiet (Quiet (..))
+
+import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
+import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
+
+data TimeInterpreterLog = MsgInterpreterPastHorizon
+  (Maybe String) -- ^ Reason for why the failure should be impossible
+  StartTime
+  PastHorizonException
+  deriving (Show)
+
+-- | Number of slots in a single epoch
+newtype EpochLength = EpochLength { unEpochLength :: Word32 }
+    deriving (Show, Eq, Generic)
+
+-- | Duration of a single slot.
+newtype SlotLength = SlotLength { unSlotLength :: NominalDiffTime }
+    deriving (Show, Eq, Generic)
+
+-- | Blockchain start time
+newtype StartTime = StartTime UTCTime
+    deriving (Show, Eq, Ord, Generic)
+
+-- | A @TimeInterpreter@ is a way for the wallet to run things of type @Qry a@,
+-- with a system start time as context.
+data TimeInterpreter m = forall eras. TimeInterpreter
+  { interpreter :: m (Interpreter eras)
+  , blockchainStartTime :: StartTime
+  }
+
+newtype Quantity (unit :: Symbol) a = Quantity { getQuantity :: a }
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving newtype (Bounded, Enum, Hashable)
+  deriving newtype NFData
+
+newtype Hash (tag :: Symbol) = Hash { getHash :: ByteString }
+  deriving stock (Generic, Eq, Ord)
+  deriving (Read, Show) via (Quiet (Hash tag))
+  deriving anyclass (NFData, Hashable)
+
+newtype SyncTolerance = SyncTolerance NominalDiffTime
+  deriving stock (Generic, Eq, Show)
+
+data SyncProgress
+  = Ready
+  | Syncing !(Quantity "percent" Percentage)
```

Comment:
I don't think the `Quantity "percent"` layer adds any additional clarity to the `Percentage` type. What do you think?

Reply:
I have removed `Quantity` as recommended.